database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java

```diff
@@ -140,18 +140,26 @@ public void onCancelled(DatabaseError error) {
         return mDataSnapshots;
     }
 
-    private int getIndexForKey(String key) {
-        int dataCount = size();
-        int index = 0;
-        for (int keyIndex = 0; index < dataCount; keyIndex++) {
-            String superKey = mKeySnapshots.getObject(keyIndex);
-            if (key.equals(superKey)) {
-                break;
-            } else if (mDataSnapshots.get(index).getKey().equals(superKey)) {
-                index++;
+    private int returnOrFindIndexForKey(int index, String key) {
+        int realIndex;
+        if (isKeyAtIndex(key, index)) {
+            realIndex = index;
+        } else {
+            int dataCount = size();
+            int dataIndex = 0;
+            for (int keyIndex = 0; dataIndex < dataCount; keyIndex++) {
```

Comment:
This is a very strange `for` loop.  It initializes `keyIndex` but then waits for a condition on `dataIndex`, all while incrementing `keyIndex` on each pass.

Took me a very long time to decipher.  I'd suggest converting it to something like this:

```java
int dataIndex = 0;
int keyIndex = 0;

while (dataIndex < dataCount) {
   // ... stuff

   keyIndex++;
}
```

Since the first line is `if (keyIndex == mKeySnapshots.size()) { break; }` you may even want to make the main condition `while(dataIndex < dataCount && keyIndex < mKeySnapshots.size())`

Also add comments throughout the method please!

Reply:
@samtstern Same! Another contributor who worked on my original PR wrote that for loop and it always takes me 5+ mins to get my head around it again. I thought that was just me! 😆 Anyway, your suggestions sound like great improvements!